File: frontend/src/app/shared/product-thumb.ts
import { Component, computed, input } from '@angular/core';
import { NzIconModule } from 'ng-zorro-antd/icon';

/** Užší tvar, než jaký vrací codegen (`ProductSummaryFieldsFragment`) — jen pole, která
 *  komponenta/pomocná funkce skutečně potřebují, ať se dá použít i pro `Product` (detail). */
export interface ProductThumbSource {
  photos?: readonly { thumbnailUrl: string }[] | null;
  externalImage?: { thumbnailUrl: string; attribution: string } | null;
}

/**
 * Stejné pravidlo jako {@link usesExternalImageFallback} — jestli produkt vlastní fotku nemá,
 * ale komunitní zápis o něm existuje, ukáže se aspoň externí zdroj (Open Food Facts). Atribuci
 * zdroje do řádku seznamu nevejde, proto ji volající strana shrne pod celý seznam jednou
 * (`usesExternalImageFallback` napříč `results()`), stejně jako na detailu zboží
 * (product-detail-page.html), jen s hromadnou poznámkou místo textu u každého řádku.
 */
export function usesExternalImageFallback(source: ProductThumbSource): boolean {
  return !(source.photos && source.photos.length > 0) && !!source.externalImage;
}

/** Miniatura zboží pro řádek výsledků (hledání, zápis ceny) — priorita zdroje: vlastní fotka >
 *  obrázek z Open Food Facts > zástupná ikona. Mobilní protějšek: ui/common/ProductThumb.kt. */
@Component({
  selector: 'app-product-thumb',
  imports: [NzIconModule],
  template: `
    @if (thumbnailUrl(); as url) {
      <img [src]="url" [alt]="name()" [title]="externalAttribution() ?? name()" class="thumb" />
    } @else {
      <span class="thumb thumb-placeholder" aria-hidden="true">
        <span nz-icon nzType="picture"></span>
      </span>
    }
  `,
  styles: `
    .thumb {
      width: 36px;
      height: 36px;
      border-radius: 4px;
      object-fit: cover;
      flex-shrink: 0;
    }
    .thumb-placeholder {
      display: inline-flex;
      align-items: center;
      justify-content: center;
      background: rgba(0, 0, 0, 0.04);
      color: rgba(0, 0, 0, 0.25);
      font-size: 16px;
    }
  `,
})
export class ProductThumb {
  readonly name = input<string>('');
  readonly photos = input<ProductThumbSource['photos']>(null);
  readonly externalImage = input<ProductThumbSource['externalImage']>(null);

  protected readonly thumbnailUrl = computed<string | null>(() => {
    const photos = this.photos();
    if (photos && photos.length > 0) return photos[0].thumbnailUrl;
    return this.externalImage()?.thumbnailUrl ?? null;
  });

  protected readonly externalAttribution = computed<string | null>(() => {
    const photos = this.photos();
    if (photos && photos.length > 0) return null;
    return this.externalImage()?.attribution ?? null;
  });
}

/** Užší tvar než {@link ProductThumbSource} — velký náhled bere plnou fotku, ne miniaturu
 *  (na plné šířce by byl thumbnail rozmazaný), takže potřebuje `url`. */
export interface ProductPreviewSource {
  photos?: readonly { url: string }[] | null;
  externalImage?: { url: string; attribution: string } | null;
}

/**
 * Velký náhled zboží nad názvem (zápis ceny) — stejná priorita zdroje jako {@link ProductThumb}
 * (vlastní fotka > Open Food Facts), ale plná fotka a `object-fit: contain`, ať se neořezávají
 * obaly s různým poměrem stran. Bez obrázku nevykreslí NIC (na rozdíl od miniatury v řádku
 * seznamu, kde zástupná ikona drží zarovnání). Mobilní protějšek: ui/common/ProductThumb.kt.
 */
@Component({
  selector: 'app-product-preview',
  template: `
    @if (imageUrl(); as url) {
      <img [src]="url" [alt]="name()" class="preview" />
    }
  `,
  styles: `
    .preview {
      display: block;
      width: 100%;
      max-height: 200px;
      object-fit: contain;
      border-radius: 4px;
      background: rgba(0, 0, 0, 0.04);
    }
  `,
})
export class ProductPreview {
  readonly name = input<string>('');
  readonly photos = input<ProductPreviewSource['photos']>(null);
  readonly externalImage = input<ProductPreviewSource['externalImage']>(null);

  protected readonly imageUrl = computed<string | null>(() => {
    const photos = this.photos();
    if (photos && photos.length > 0) return photos[0].url;
    return this.externalImage()?.url ?? null;
  });
}

File: mobile/app/src/main/java/cz/kvalitacena/ui/common/ProductThumb.kt
package cz.kvalitacena.ui.common

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.res.painterResource
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import coil3.compose.AsyncImage
import cz.kvalitacena.R
import cz.kvalitacena.network.ExternalProductImage
import cz.kvalitacena.network.Photo

/**
 * Priorita zdroje obrázku zboží: vlastní fotka > obrázek z Open Food Facts > žádný. Jedno
 * místo pro [ProductThumb] i [ProductImagePreview], stejné pravidlo jako na detailu zboží
 * (ProductDetailScreen.kt) a na webu (shared/product-thumb.ts).
 *
 * `full = true` vrací plnou verzi (velký náhled), jinak miniaturu do řádku seznamu.
 * `externalImage.url`/`thumbnailUrl` jsou (na rozdíl od [Photo.fullUrl]/[Photo.thumbUrl]) už
 * plné externí URL — OFF obrázky se neukládají do core.media, appka je nikdy neprefixuje
 * ApiConfig.BASE_URL.
 */
fun productImageUrl(photos: List<Photo>, externalImage: ExternalProductImage?, full: Boolean = false): String? {
  val photo = photos.firstOrNull()
  if (photo != null) return if (full) photo.fullUrl() else photo.thumbUrl()
  return if (full) externalImage?.url else externalImage?.thumbnailUrl
}

/**
 * Zobrazený obrázek pochází z Open Food Facts, ne z vlastní fotky — volající MUSÍ vedle něj
 * uvést atribuci zdroje (ODbL, docs/datovy-model.md). Web protějšek:
 * `usesExternalImageFallback` v shared/product-thumb.ts.
 */
fun usesExternalImage(photos: List<Photo>, externalImage: ExternalProductImage?): Boolean =
  photos.isEmpty() && externalImage != null

/**
 * Miniatura zboží pro řádek výsledků hledání — priorita zdroje [productImageUrl], bez obrázku
 * zástupná ikona (drží zarovnání řádků). Web protějšek: shared/product-thumb.ts.
 */
@Composable
fun ProductThumb(
  name: String,
  photos: List<Photo>,
  externalImage: ExternalProductImage?,
  modifier: Modifier = Modifier,
  size: Dp = 36.dp,
) {
  val url = productImageUrl(photos, externalImage)

  if (url != null) {
    AsyncImage(
      model = url,
      contentDescription = name,
      contentScale = ContentScale.Crop,
      modifier = modifier.size(size).clip(RoundedCornerShape(12.dp)),
    )
  } else {
    Box(
      modifier = modifier
        .size(size)
        .clip(RoundedCornerShape(12.dp))
        .background(MaterialTheme.colorScheme.surfaceVariant),
      contentAlignment = Alignment.Center,
    ) {
      Icon(
        painter = painterResource(R.drawable.ic_picture),
        contentDescription = null,
        tint = MaterialTheme.colorScheme.onSurfaceVariant,
        modifier = Modifier.size(size * 0.5f),
      )
    }
  }
}

/**
 * Velký náhled zboží nad názvem (obrazovka zápisu ceny po skenu) — stejná priorita zdroje jako
 * [ProductThumb], ale plná URL místo miniatury (na plné šířce by byl thumbnail rozmazaný) a
 * `ContentScale.Fit`, ať se neořezávají obaly s různým poměrem stran.
 *
 * Bez obrázku nevykreslí NIC (na rozdíl od [ProductThumb]) — velký prázdný rámeček s ikonou by
 * jako hlavní prvek obrazovky jen odsunul ceny pod ohyb.
 */
@Composable
fun ProductImagePreview(
  name: String,
  photos: List<Photo>,
  externalImage: ExternalProductImage?,
  modifier: Modifier = Modifier,
  height: Dp = 180.dp,
) {
  val url = productImageUrl(photos, externalImage, full = true) ?: return

  AsyncImage(
    model = url,
    contentDescription = name,
    contentScale = ContentScale.Fit,
    modifier = modifier
      .height(height)
      .clip(RoundedCornerShape(12.dp))
      .background(MaterialTheme.colorScheme.surfaceVariant),
  )
}
